sklearn/utils/stats.py

```diff
@@ -66,4 +68,18 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     col_index = np.arange(array.shape[1])
     percentile_in_sorted = sorted_idx[percentile_idx, col_index]
     percentile = array[percentile_in_sorted, col_index]
+
+    # percentiles that point to nan values are redirected to the next lower value:
+    while bool(np.isnan(percentile).any()):
+        percentile_idx[np.isnan(percentile)] -= 1
+        try:
+            percentile_in_sorted = sorted_idx[percentile_idx, col_index]
+        except IndexError:
+            raise ValueError(
+                "One feature contains too many NaN values. This error should "
+                "actually either raise within the API, or there needs to be some "
+                "validation here before to make sure it cannot happen."
+            )
```

Comment:
I think it's fine to return `np.nan` valued percentile (maybe with a warning?) if some array is all-nan valued.

Reply:
The reason I had put it here was that I had tested this on the branch from the [SplineTransformer PR](https://github.com/scikit-learn/scikit-learn/pull/28043) where we add `test_spline_transformer_handles_all_nans` (quite a special case anyway) and this made it raise in a not very informative way. 

I think the error needs to be handled somehow, but within `SplineTransformer`, so that `_weighted_percentile` can stay flexible in case it's used differently in another context at one point.

I will remove it from here.

Reply:
I have looked at this again and went through all the places where `_weighted_percentile` gets used.

The losses I mentioned above (only internal use) as well as `set_huber_delta()` and `median_absolute_error()` find a percentile from within `y_true` or `y_true-some_other_1darray` , where I think it should be pretty obvious for the user if all values were nan. This is why I think it's okay without a warning.

`KBinsDiscretizer` and `SplineTransformer` use it on the whole `X` within their fit methods and I think we could add a warning there for all-nan-columns if this kind of validation is not already done somewhere else.

Reply:
When `_weighted_percentile` is called by the loss functions, we can expect that there will never be any nan values in `y_true` because the public method of the estimator (typically `.fit`) always checks that there are no nan values and raise an exception if needed before calling `_weighted_percentile`:

```python
>>> from sklearn.datasets import make_regression
>>> import numpy as np
>>> from sklearn.linear_model import QuantileRegressor
>>> X, y = make_regression()
>>> y[0] = np.nan
>>> QuantileRegressor().fit(X, y)
Traceback (most recent call last):
  Cell In[14], line 1
    QuantileRegressor().fit(X, y)
  File ~/code/scikit-learn/sklearn/base.py:1514 in wrapper
    return fit_method(estimator, *args, **kwargs)
  File ~/code/scikit-learn/sklearn/linear_model/_quantile.py:163 in fit
    X, y = self._validate_data(
  File ~/code/scikit-learn/sklearn/base.py:650 in _validate_data
    X, y = check_X_y(X, y, **check_params)
  File ~/code/scikit-learn/sklearn/utils/validation.py:1286 in check_X_y
    y = _check_y(y, multi_output=multi_output, y_numeric=y_numeric, estimator=estimator)
  File ~/code/scikit-learn/sklearn/utils/validation.py:1308 in _check_y
    _assert_all_finite(y, input_name="y", estimator_name=estimator_name)
  File ~/code/scikit-learn/sklearn/utils/validation.py:123 in _assert_all_finite
    _assert_all_finite_element_wise(
  File ~/code/scikit-learn/sklearn/utils/validation.py:172 in _assert_all_finite_element_wise
    raise ValueError(msg_err)
ValueError: Input y contains NaN.
```

and similarly for public scoring functions that rely on those losses:

```python
>>> from sklearn.metrics import mean_pinball_loss
>>> mean_pinball_loss(y, y)
Traceback (most recent call last):
  Cell In[16], line 1
    mean_pinball_loss(y, y)
  File ~/code/scikit-learn/sklearn/utils/_param_validation.py:213 in wrapper
    return func(*args, **kwargs)
  File ~/code/scikit-learn/sklearn/metrics/_regression.py:318 in mean_pinball_loss
    y_type, y_true, y_pred, multioutput = _check_reg_targets(
  File ~/code/scikit-learn/sklearn/metrics/_regression.py:112 in _check_reg_targets
    y_true = check_array(y_true, ensure_2d=False, dtype=dtype)
  File ~/code/scikit-learn/sklearn/utils/validation.py:1056 in check_array
    _assert_all_finite(
  File ~/code/scikit-learn/sklearn/utils/validation.py:123 in _assert_all_finite
    _assert_all_finite_element_wise(
  File ~/code/scikit-learn/sklearn/utils/validation.py:172 in _assert_all_finite_element_wise
    raise ValueError(msg_err)
ValueError: Input contains NaN.
```

Still we might want to add a `nan_policy` parameter to `_weighted_percentile` with `"raise"` and `"omit"` as possible options as scipy does:

- https://github.com/scipy/scipy/blob/main/doc/source/dev/api-dev/nan_policy.rst#L70

This would make the use of `_weighted_percentile` within or code base more explicit: loss functions would call it with `nan_policy="error"` (the default) while transformers such as `SplineTransformer` that actually need nan support would call it with `nan_policy="omit"`.

WDYT?

Reply:
I think that the occasion when we would want to raise (all values from a column being nans) is so edge, that we should not have a param in `_weighted_percentile` for it.

Having something like scipys `nan_policy` only for this purpose seems like an overkill for me, also because `np.nanpercentile` will be the future in a determined time anyways.

For the other cases, when there are only a few nan values, the users can express their intend in `SplineTransformer.fit()` with `handle_missing='zeros'` (after #28043 is merged) where `_weighted_percentile` is only called if the user explicitly wants to deal with nans. So there would be no reason to deselect `_weighted_percentile` nan-dealing behaviour.

I wonder if for `KBinsDiscretizer`,  which is a special case of `SplineTransformer`, we want to do something similar or test and document what changes when `_weighted_percentile` supports nans (right now all tests pass, which means it is not tested with nans). This would be a new PR for me.

Reply:
> Having something like scipys nan_policy would be a larger design decision for the whole project and having it for our little purpose seems like an overkill to me, also because np.nanpercentile will be the future in a determined time anyways.

If we introduce this just for private methods like `_weighted_percentile`, then it's fine. But indeed, this kind errors are better be raised by public methods because there are more likely to be able to output error messages with more informative contextual details than generic private method would.

So ok for not introducing `nan_policy` as part of this PR.

Reply:
Thank you, I'm relieved. And sorry if I expressed myself a bit strong there.